src/librustc_mir/build/expr/stmt.rs

```diff
@@ -79,14 +79,31 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 block.unit()
             }
             ExprKind::Continue { label } => {
-                this.break_or_continue(expr_span, label, block,
-                                       |loop_scope| loop_scope.continue_block)
+                let LoopScope { continue_block, extent, .. } =
+                    *this.find_loop_scope(expr_span, label);
+                this.break_or_continue(expr_span, block, continue_block, extent)
             }
-            ExprKind::Break { label } => {
-                this.break_or_continue(expr_span, label, block, |loop_scope| {
-                    loop_scope.might_break = true;
-                    loop_scope.break_block
-                })
+            ExprKind::Break { label, mut value } => {
+                let (break_block, extent, destination) = {
+                    let LoopScope {
+                        ref mut might_break,
+                        break_block,
+                        extent,
+                        ref destination,
+                        ..
+                    } = *this.find_loop_scope(expr_span, label);
+                    *might_break = true;
+                    // FIXME unnecessary clone
+                    (break_block, extent, destination.clone())
+                };
+                if let Some(dest) = destination {
+                    if let Some(value) = value.take() {
```

Comment:
> Also, the first branch already takes care of the returned block (it uses unpack!

Ah, I missed the `block = ` inside the `unpack`.

> The reason why this code is so awkward is that we can't keep this borrowed

You can manually relinquish the borrow by doing something along the lines of:

```
let (extent, break_block) = {
let scope = this.find_loop_scope(expr_span, label);
scope.might_break = true;
if let Some(ref dst) = scope.destination {
    block /* this is important! */ = match value {
        Some(value) => unpack!(block = this.into(dst, block, value)),
        None => this.cfg.push_assign_unit(block, source_info, dst),
    };
}
(scope.extent, scope.break_block)
};
this.exit_scope(expr_span, extent, break_block, block);
this.cfg.start_new_block().unit()
```

This should work because both `Extent` and `BasicBlock` are Copy AFAIR.